### DISCUSSION TITLE
Masking Field Value - Subscriptions/event listeners for successful reset/setValue/etc.(?)

### DISCUSSION THREAD

**Discussion by KershSoftware** [react-hook-form/react-hook-form#8185] [Q&A] [answered]
So our team was working towards applying an existing react masking library (https://github.com/uNmAnNeR/imaskjs/tree/master/packages/react-imask) along with our usage of react-hook-form.

The mask works just fine across the board in terms of correctly working with our existing component library and how we've integrated our library with react-hook-form either through the register method or the Controller wrapper component.

However, we discovered a particularly specific and annoying issue. It's actually clear to me what is wrong but rather not clear how to solve the problem. **When the form field is programmatically set with `setValue` or the entire form is reset with `reset`, the mask does not trigger.** Additionally, and actually more precisely the bigger bug: **The user typing into the field after `reset` or `setValue` causes excess values to be injected into the form field due to the mask being out of sync.** (in the codepen below: try clicking the Set to 9999.99 button, click on the 9 before the decimal and type a value, you'll see what I mean)

I've attached a working example of the bug below:
https://codesandbox.io/s/react-hook-form-with-react-imask-bug-example-bks8o6?file=/src/MaskedInput.tsx

Now I'm aware the mask issue is not inherently an issue with `react-hook-form`, but rather I'm posting here to ask for advice for anyone who has achieved this sort of masking. I'm a fan of the `imask` library specifically as it satisfies our masking requirements really well.

You'll notice the console spits out the following when `setValue` or `reset` fires and focus is brought to the form field:
```
Element value was changed outside of mask. Syncronize mask using `mask.updateValue()` to work properly. 
```

Clearly there is a way to programmatically set the mask as well given this method; but this leads to the title of my post -- I can't just watch for the form field value within the FormProvider's context to change and always trigger the mask (because the mask in turn sets the form context value). Which got me wondering if there was a way to subscribe to a `reset` or a `setValue` successfully completing? Or any other possible ideas anyone may have?

Thanks for reading and any potential help!

**Comment by leapful** [comment, marked as answer]
@Kersheh : It's a little bit tricky when working with imask because it also handles input data by using ref element.

In your example, when using "setValue" of "react-hook-from", the value must be synchronized with "react-imask", but when typing new value on input (which is controlled by "react-imask"), it must be synchronized back to "react-hook-form".

I would suggest to use only handling from "react-imask" on the input element and then sync data between it with "react-hook-form".

Demo: https://codesandbox.io/s/react-hook-form-with-react-imask-bug-example-forked-c064hk?file=/src/MaskedInput.tsx

**Reply by KershSoftware** [reply]
Fantastic, seems to be working as expected now. However, I did notice one gap that may introduce bugs later on. `touched` and `dirty` are tied to the Controller's `onChange` and `onBlur`. I can reintroduce `onBlur` again for the `touched` state to properly update. However, I cannot reintroduce the Controller's render `onChange` as it defeats the purpose of imask taking control.

I can add `{ shouldDirty: true }` to the `setValue` called inside of the imask `onComplete`; however, this always makes the field `dirty` because this gets call on initialization of the form.

**Reply by KershSoftware** [reply]
Maybe it makes sense to add some custom guard to only set `shouldDirty` to `true` if the value to be set differs form the forms default value for that field by name.

Edit: Yep a simple guard comparing against the default value does the trick on my end.

**Comment by npearson72** [comment]
I realize this is a slightly dated question. I ran into a similar situation but recently upgraded my version of react imask, and was able to get things working in a much simpler way. Here's an example based off your original sandbox: https://codesandbox.io/s/react-hook-form-with-react-imask-bug-example-forked-rcnpzu?file=/src/MaskedInput.tsx

Hope this helps
